pennylane/_qubit_device.py

```diff
@@ -2011,3 +2013,27 @@ def _adjoint_jacobian_processing(jac):
 
         # must be 2-dimensional
         return tuple(tuple(np.array(j_) for j_ in j) for j in jac)
+
+    def _get_diagonalizing_gates(self, circuit: QuantumTape):  # pylint:disable=no-self-use
+        """Returns the gates that diagonalize the measured wires such that they
+        are in the eigenbasis of the circuit observables.
+
+        Note that this exists as a method of the Device class to enable child classes to
+        override the implementation if necessary (for example, to skip computing rotation
+        gates for a measurement that doesn't need them).
+
+        Args:
+            circuit (~.tape.QuantumTape): The circuit containing observables that may need diagonalizing
+
+        Returns:
+            List[~.Operation]: the operations that diagonalize the observables
+        """
+        rotation_gates = []
+
+        with qml.queuing.QueuingManager.stop_recording():
+            for observable in circuit.observables:
+                # some observables do not have diagonalizing gates,
+                # in which case we just don't append any
+                with contextlib.suppress(qml.operation.DiagGatesUndefinedError):
+                    rotation_gates.extend(observable.diagonalizing_gates())
+        return rotation_gates
```

Comment:
that was my thought process, but I'm ok to do that since this is just default behaviour. And if someone wants to remove it down the road, they can of course copy-paste the contents to this method (they won't miss it - tests would fail)